// @s-ui/react-atom-button/lib/Button.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
var _excluded = ["children", "href", "target", "disabled", "isSubmit", "isButton", "link", "linkFactory", "forwardingRef"],
    _excluded2 = ["children"];
import PropTypes from 'prop-types';
import { forwardRef } from 'react';
import { jsx as _jsx } from "react/jsx-runtime";

var Button = function Button(_ref) {
  var children = _ref.children,
      href = _ref.href,
      target = _ref.target,
      disabled = _ref.disabled,
      isSubmit = _ref.isSubmit,
      isButton = _ref.isButton,
      link = _ref.link,
      Link = _ref.linkFactory,
      forwardingRef = _ref.forwardingRef,
      attrs = _objectWithoutPropertiesLoose(_ref, _excluded);

  if (isSubmit) attrs.type = 'submit';
  if (isButton) attrs.type = 'button';
  var defaultRel = target === '_blank' ? 'noopener' : undefined;
  var rel = attrs.rel || defaultRel;
  return link ? /*#__PURE__*/_jsx(Link, _extends({}, attrs, {
    href: href,
    target: target,
    rel: rel,
    ref: forwardingRef,
    children: children
  })) : /*#__PURE__*/_jsx("button", _extends({}, attrs, {
    disabled: disabled,
    ref: forwardingRef,
    children: children
  }));
};

Button.defaultProps = {
  // eslint-disable-next-line react/prop-types
  linkFactory: /*#__PURE__*/forwardRef(function (_ref2, ref) {
    var children = _ref2.children,
        props = _objectWithoutPropertiesLoose(_ref2, _excluded2);

    return /*#__PURE__*/_jsx("a", _extends({}, props, {
      ref: ref,
      children: children
    }));
  })
};
export default Button;
